mailbox/tool/src/main/java/org/apache/james/mailbox/indexer/ReIndexerImpl.java

```diff
@@ -0,0 +1,138 @@
+ *  mails for instance, you will see it in the indexed value !
+ *
+ *  Why only care about updates and deletions ? Additions are already handled by the indexer that behaves normaly. We
+ *  should just "adapt" our indexed value to the latest value, if any. The normal indexer will take care of new stuff.
+ */
+public class ReIndexerImpl<Id extends MailboxId> implements ReIndexer {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ReIndexerImpl.class);
+    public static final int LIMIT = 0;
+
+    private final MailboxManager mailboxManager;
+    private final ListeningMessageSearchIndex<Id> messageSearchIndex;
+    private final MailboxSessionMapperFactory<Id> mailboxSessionMapperFactory;
+
+    public ReIndexerImpl(MailboxManager mailboxManager,
+                         ListeningMessageSearchIndex<Id> messageSearchIndex,
+                         MailboxSessionMapperFactory<Id> mailboxSessionMapperFactory) {
+        this.mailboxManager = mailboxManager;
+        this.messageSearchIndex = messageSearchIndex;
+        this.mailboxSessionMapperFactory = mailboxSessionMapperFactory;
+    }
+
+    public void reIndex(MailboxPath path) throws MailboxException {
+        MailboxSession mailboxSession = mailboxManager.createSystemSession("re-indexing", LOGGER);
+        reIndex(path, mailboxSession);
+    }
+
+
+    public synchronized void reIndex() throws MailboxException {
+        MailboxSession mailboxSession = mailboxManager.createSystemSession("re-indexing", LOGGER);
+        List<MailboxPath> mailboxPaths = mailboxManager.list(mailboxSession);
+        GlobalRegistration globalRegistration = new GlobalRegistration();
+        mailboxManager.addGlobalListener(globalRegistration, mailboxSession);
+        for (MailboxPath mailboxPath : mailboxPaths) {
+            if (globalRegistration.indexThisPath(mailboxPath)) {
+                reIndex(mailboxPath, mailboxSession);
+            }
+        }
+        mailboxManager.removeGlobalListener(globalRegistration, mailboxSession);
+    }
+
+    private void reIndex(MailboxPath path, MailboxSession mailboxSession) throws MailboxException {
+        MailboxRegistration mailboxRegistration = new MailboxRegistration(path);
+        mailboxManager.addListener(path, mailboxRegistration, mailboxSession);
+        LOGGER.info("Intend to reindex " + path);
+        Mailbox<Id> mailbox = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).findMailboxByPath(path);
+        messageSearchIndex.delete(mailboxSession, mailbox, MessageRange.all());
+        handleIterations(mailboxSession,
+            mailboxRegistration,
+            mailbox,
+            mailboxSessionMapperFactory.getMessageMapper(mailboxSession)
+                .findInMailbox(mailbox,
+                    MessageRange.all(),
+                    MessageMapper.FetchType.Full,
+                    LIMIT));
+        mailboxManager.removeListener(path, mailboxRegistration, mailboxSession);
```

Comment:
Thanks for this! If a single document can not be indexed, the process stops and we leak the listener. I guess the wanted behaviour is to not throw during a single docurement indexation, and ensure we remove the listener.